### DISCUSSION TITLE
Concurrent execution, potential bug

### DISCUSSION THREAD

**Discussion by IlyaVassyutovich** [quartznet/quartznet#1921] [Q&A]
Hello!

I've tried to setup Quartz with cluster configuration with two instances of my app running on separate machines/k8s-pods. And I need to run periodical job, but only in one instance. So I've used a code like this:

```csharp
quartzConfig.ScheduleJob<TestQuartzJob>(
	trigger => trigger
		.StartNow()
		.WithIdentity($"{nameof(TestQuartzJob)}_trigger")
		.WithSimpleSchedule(s => s.WithInterval(TimeSpan.FromSeconds(30)).RepeatForever()),
	job => job
		.WithIdentity($"{nameof(TestQuartzJob)}_job")
		.DisallowConcurrentExecution()); // <<< here
```

But when I run two instances of my app, I have jobs executing simultaneously on both machines. But when I mark my `TestQuartzJob`-class with `DisallowConcurrentExecutionAttribute` everything start working as expected — jobs run in non-cocurrent mode, only single instance of a job per time.

-----

I think I found a cause for this behavior — [here in StdAdoDelegate.Jobs.cs](https://github.com/quartznet/quartznet/blob/main/src/Quartz/Impl/AdoJobStore/StdAdoDelegate.Jobs.cs#L132-L137) we materialize `JobDetail` without reading `IS_NONCONCURRENT` from DB and without configuring JobBuilder with concucrrency disallowed flag.

**Comment by lahma** [comment]
Thanks for the report and detailed analysis! It indeed seems wrong that the `IS_NONCONCURRENT` isn't being handled here. I think it originates from the fact that this information was only allowed to be read from attribute information and later came the possibility to also change it via code.

Would you like to offer a PR against the 3.x branch, one option is to add `internal` setter for the property. The main uses job builder and 3.x the "old way", both of course missing this bit. If you are really into helping out, two PRs against both branches would be most welcome!

**Reply by IlyaVassyutovich** [reply]
Yeah, I've already almost done with default branch. Will investigate code in 3.x branch and prepare PR for this branch also.

**Reply by IlyaVassyutovich** [reply]
Here comes [PR](https://github.com/quartznet/quartznet/pull/1922) for `main`.

**Reply by IlyaVassyutovich** [reply]
Here comes [another one](https://github.com/quartznet/quartznet/pull/1923).

**Comment by IlyaVassyutovich** [comment]
@lahma Good sir, can you tell me when to expect new 3.x release with fixes for concurrency?

**Reply by lahma** [reply]
@IlyaVassyutovich 3.6.1 should now be available on NuGet, thanks for contributing to the release!

**Reply by IlyaVassyutovich** [reply]
Thank you so much!

**Comment by IlyaVassyutovich** [comment]
@lahma Hello again 😊 

I guess I've found another bug in concurrency implementation, when it's configured with code, not with attribute.

Previously, before my changes in `SelectJobDetail` concurrency restrictions were completely ignored. That is, job were starting and running in parallel all the time while my app was running.

My fix with correct materialization helped... Almost. Now, if I run two instances of my app, synchronisation between jobs work only after first execution. I'll explain via example:

```
[instance-0]: app started
[instance-0]: scheduler started
[instance-0]: Test job with key "ABCDEF" started # it's long running so it will be in "execution"-state for some time
[instance-1]: app started
[instance-1]: scheduler started
[instance-1]: Test job with key "ABCDEF" started # it's running concurrently with job from instance-0
[instance-0]: Test job with key "ABCDEF" completed # we'll wait for some time after that
[instance-1]: Test job with key "ABCDEF" completed
[instance-0]: Test job with key "ABCDEF" started
# after first job completion in instance-1 app will "respect" concurrency restrictions and will not run while job from other instance in progress
```

Also, here are logs collected from Quartz — [without attribute](https://gist.github.com/IlyaVassyutovich/3023123749c6c8fdcf220e7c49b666d9).

I've tried to pinpoint the source of such behaviour, but was not successful, maybe you can help me find this bug.

I will try to prepare repro project with SqlServer in docker tomorrow.

**Reply by IlyaVassyutovich** [reply]
@lahma Hello there!

Here's repro https://github.com/IlyaVassyutovich/quartz-tests. I've tried my best to be as concise as possible with minimal configuration. But the database requirements...

Also, [here's](https://gist.github.com/IlyaVassyutovich/91a17a81793c30a536bd49171a2c0bdf) log from my run, that shows the problem with concurrent running of jobs (frist run).

**Reply by IlyaVassyutovich** [reply]
@lahma Hello, friend! Any news regarding this issue?

**Reply by lahma** [reply]
Nope, no news.

**Reply by IlyaVassyutovich** [reply]
https://github.com/quartznet/quartznet/issues/2011
